server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java

```diff
@@ -321,13 +330,23 @@ private String getRemoteAddr() {
         }
     }
 
-    private List<MailAddress> getSenderAsList(Mail originalMail) {
-        MaybeSender reversePath = originalMail.getMaybeSender();
+    private List<MailAddress> getSenderAsList(Mail originalMail) throws MessagingException {
+        Optional<MailAddress> returnPath = Optional.ofNullable(originalMail.getMessage().getHeader(RETURN_PATH))
+                .flatMap(path -> path.length > 0 ? Optional.ofNullable(path[0]) : Optional.empty())
+                .map((Throwing.<String, MailAddress>function(MailAddress::new).sneakyThrow()));
 
-        if (getInitParameters().isDebug()) {
-            LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", reversePath.asString());
+        if(returnPath.isPresent()) {
+            if (getInitParameters().isDebug()) {
+                LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", returnPath.get().asString());
+            }
+            return ImmutableList.of(returnPath.get());
+        } else {
+            MaybeSender reversePath = originalMail.getMaybeSender();
```

Comment:
Maybe we could falling back to To MIME fields before falling back to the SMTP envelope sender?

Reply:
are you sure it makes sense for a DSN? (I really don't know, but would be happy to read some clue about this)

Reply:
(no idea: I'm not a DSN expert)

Reply:
Actually https://tools.ietf.org/html/rfc3461 states:

```
   The DSN MUST be addressed (in both the message header and the
   transport envelope) to the return address from the transport envelope
   which accompanied the original message for which the DSN was
   generated.  (For a message that arrived via SMTP, the envelope return
   address appears in the MAIL FROM command.)
```

:-/

Reply:
The quote is taken from https://tools.ietf.org/html/rfc3464 `An Extensible Message Format for Delivery Status Notifications` https://tools.ietf.org/html/rfc3464#section-2

(Sorry I messed up)

Reply:
https://tools.ietf.org/html/rfc3834 `Recommendations for Automatic Responses to Electronic Mail`

```
In general, automatic responses SHOULD be sent to the Return-Path
   field if generated after delivery.  If the response is generated
   prior to delivery, the response SHOULD be sent to the reverse-path
   from the SMTP MAIL FROM command, or (in a non-SMTP system) to the
   envelope return address which serves as the destination for non-
   delivery reports.
```